source/Halibut/Transport/PollingClient.cs

```diff
@@ -13,47 +15,62 @@ public class PollingClient : IPollingClient
-            : this(subscription, secureClient, handleIncomingRequest, null, createRetryPolicy)
-        {
-        }
 
-        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, ILog log, Func<RetryPolicy> createRetryPolicy)
-            : this(subscription, secureClient, handleIncomingRequest, log, CancellationToken.None, createRetryPolicy)
-        {
-        }
+        Task? pollingClientLoopTask;
+        readonly CancellationTokenSource workingCancellationTokenSource;
+        readonly CancellationToken cancellationToken;
+        
+        readonly Func<RetryPolicy> createRetryPolicy;
+        readonly AsyncHalibutFeature asyncHalibutFeature;
+        
 
-        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, ILog log, CancellationToken cancellationToken, Func<RetryPolicy> createRetryPolicy)
+        public PollingClient(Uri subscription, ISecureClient secureClient, Func<RequestMessage, ResponseMessage> handleIncomingRequest, ILog log, CancellationToken cancellationToken, Func<RetryPolicy> createRetryPolicy, AsyncHalibutFeature asyncHalibutFeature)
         {
             this.subscription = subscription;
             this.secureClient = secureClient;
             this.handleIncomingRequest = handleIncomingRequest;
             this.log = log;
             this.cancellationToken = cancellationToken;
+            workingCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
             this.createRetryPolicy = createRetryPolicy;
-            thread = new Thread(ExecutePollingLoop);
-            thread.Name = "Polling client for " + secureClient.ServiceEndpoint + " for subscription " + subscription;
-            thread.IsBackground = true;
+            this.asyncHalibutFeature = asyncHalibutFeature;
+            
         }
 
         public void Start()
         {
             working = true;
-            thread.Start();
+
+            if (asyncHalibutFeature == AsyncHalibutFeature.Disabled)
+            {
+                pollingClientLoopThread = new Thread(ExecutePollingLoop!);
+                pollingClientLoopThread.Name = "Polling client for " + secureClient.ServiceEndpoint + " for subscription " + subscription;
+                pollingClientLoopThread.IsBackground = true;
+                pollingClientLoopThread.Start();
+            }
+            else
+            {
+                pollingClientLoopTask = Task.Run(async () => await ExecutePollingLoopAsyncCatchingExceptions(workingCancellationTokenSource.Token));
+            }
         }
 
         public void Dispose()
         {
             working = false;
-        }
+            workingCancellationTokenSource.Cancel();
```

Comment:
I think this try/catch is the wrong way around.

I.e., you can call 'Dispose' as much as you want, but 'Cancel' will throw if it is already disposed.